TESTS/Level2/Pwm/Pwm.cpp

```diff
@@ -0,0 +1,164 @@
+    TEST_ASSERT_MESSAGE( std::abs(rc-fc) <= calculated_percent, "There was more than a specific variance in number of rise vs fall cycles");
+    TEST_ASSERT_MESSAGE( std::abs(iterations - rc) <= calculated_percent, "There was more than a specific variance in number of rise cycles seen and number expected.");
+    TEST_ASSERT_MESSAGE( std::abs(iterations - fc) <= calculated_percent, "There was more than a specific variance in number of fall cycles seen and number expected.");
+    // TEST_ASSERT_MESSAGE( std::abs(expectedTime - avgTime) <= calculated_percent,"Greater than a specific variance between expected and measured duty cycle");
+
+}
+
+// Test case to just iterate through duty cycles
+void test_pwm_dutycycle(PinName pin, float tolerance, int iterations) {
+	DEBUG_PRINTF("Running pwm test on pin %d\n", pin);
+    TEST_ASSERT_MESSAGE(pin != NC, "Pin is NC");
+    
+	for (float dutycycle=0.05f; dutycycle <= 0.95f; dutycycle+=0.1f) {
+		test_pwm_execute(pin, tolerance, iterations, dutycycle, 10);
+	}
+
+}
+
+// Test case to just iterate through the period
+void test_pwm_frequency(PinName pin, float tolerance, int iterations) {
+	DEBUG_PRINTF("Running pwm test on pin %d\n", pin);
+    TEST_ASSERT_MESSAGE(pin != NC, "Pin is NC");
+
+    for (int period=10; period<=200; period+=40) {
+		test_pwm_execute(pin, tolerance, iterations, 0.5f, period);
+	}
+}
+
+// Test case that combines period and duty cycles
+void test_pwm(PinName pin, float tolerance, int iterations) {
+	DEBUG_PRINTF("Running pwm test on pin %d\n", pin);
+    TEST_ASSERT_MESSAGE(pin != NC, "Pin is NC");
+
+	for (float dutycycle=0.2f; dutycycle <= 0.8f; dutycycle+=0.2f) {
+    	for (int period=10; period<=210; period+=40) {
+			test_pwm_execute(pin, tolerance, iterations, dutycycle, period);
+		}
+	}
+}
+
+utest::v1::control_t test_level2_pwm_frequency(const size_t call_count) {
+	return TestFramework::test_level2_framework(TestFramework::PWM, TestFramework::CITS_PWM, &test_pwm_frequency, 0.05f, 20);
+}
+
+utest::v1::control_t test_level2_pwm_dutycycle(const size_t call_count) {
+	return TestFramework::test_level2_framework(TestFramework::PWM, TestFramework::CITS_PWM, &test_pwm_dutycycle, 0.05f, 20);
+}
+
+utest::v1::control_t test_level2_pwm(const size_t call_count) {
+	return TestFramework::test_level2_framework(TestFramework::PWM, TestFramework::CITS_PWM, &test_pwm, 0.05f, 20);
+}
+
+Case cases[] = {
+	Case("Level 2 - PWM Frequency test (all pins)", test_level2_pwm_frequency, TestFramework::greentea_failure_handler),
+	Case("Level 2 - PWM Dutycycle test (all pins)", test_level2_pwm_dutycycle, TestFramework::greentea_failure_handler),
+	Case("Level 2 - PWM all tests (all pins)", test_level2_pwm, TestFramework::greentea_failure_handler),
+};
+
+int main() {
+	// Formulate a specification and run the tests based on the Case array
+	Specification specification(TestFramework::test_setup<200>, cases);
+    return !Harness::run(specification);
```

Comment:
I can move to 300s. I tested with the K64F and Nucleo_F429ZI in which 200s was enough. Forgot that additional targets will have more PWMs which will increase the time